### ISSUE TITLE
Replicaset connection issue

### ISSUE THREAD

**Issue by cbratschi** [christkv/mongodb-core#163]
While upgrading from Mongoose 4.7.8 to 4.8 we suddenly had problems to connect to our MongoDB replica set. After analyzing the modified files we discovered that replset.js is the root cause. If we undo the lastest changes it works again. Particularly those two changes:

https://github.com/christkv/mongodb-core/commit/94d52180a03b42d40cee2e7cdc396b6f5d275593

https://github.com/christkv/mongodb-core/commit/9cf1dd46595bcaf5b9ef5227bbdd679247def750

Starting from Mongoose 4.7.9 it depends on the latest mongodb-core version.

I guess it is due to the async handling introduced by the timers which causes problems in the state handling of Mongoose. Probably the connection is reported too early as established.

Errors seen while using the latest Mongoose version:

1) promise errors

```
(node:6384) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 40): TypeError: Cannot read property 'hasListCollectionsCommand' of null
(node:6384) DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
(node:6384) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 41): TypeError: Cannot read property 'hasListCollectionsCommand' of null
```

2) crash

```
Unhandled rejection MongoError: cannot connect to server
    at Collection.listIndexes (Z:\git\pharos-node-server\node_modules\mongoose\node_modules\mongodb\lib\collection.js:1824:11)
    at indexInformation (Z:\git\pharos-node-server\node_modules\mongoose\node_modules\mongodb\lib\db.js:1670:25)
    at Db.indexInformation (Z:\git\pharos-node-server\node_modules\mongoose\node_modules\mongodb\lib\db.js:1635:44)
    at ensureIndex (Z:\git\pharos-node-server\node_modules\mongoose\node_modules\mongodb\lib\db.js:1144:8)
    at Db.ensureIndex (Z:\git\pharos-node-server\node_modules\mongoose\node_modules\mongodb\lib\db.js:1119:44)
    at ensureIndex (Z:\git\pharos-node-server\node_modules\mongoose\node_modules\mongodb\lib\collection.js:1895:13)
    at Collection.ensureIndex (Z:\git\pharos-node-server\node_modules\mongoose\node_modules\mongodb\lib\collection.js:1883:44)
    at Collection.tryCatcher (Z:\git\pharos-node-server\node_modules\bluebird\js\release\util.js:16:23)
    at Collection.ret [as ensureIndexAsync] (eval at makeNodePromisifiedEval (Z:\git\pharos-node-server\node_modules\bluebird\js\release\promisify.js:184:12), <anonymous>:13:39)
    at MongoStore.setAutoRemoveAsync (Z:\git\pharos-node-server\node_modules\connect-mongo\src\index.js:129:40)
    at MongoStore.handleNewConnectionAsync (Z:\git\pharos-node-server\node_modules\connect-mongo\src\index.js:121:18)
    at MongoStore (Z:\git\pharos-node-server\node_modules\connect-mongo\src\index.js:89:26)
    at Object.<anonymous> (Z:\git\pharos-node-server\app-main.js:121:12)
    at Module._compile (module.js:571:32)
    at Object.Module._extensions..js (module.js:580:10)
    at Module.load (module.js:488:32)
    at tryModuleLoad (module.js:447:12)
    at Function.Module._load (module.js:439:3)
    at Module.require (module.js:498:17)
    at require (internal/module.js:20:19)
    at Array.mongoose.waitForConnection.err (Z:\git\pharos-node-server\app.js:27:5)
    at NativeConnection.callback (Z:\git\pharos-node-server\models\mongoose.js:80:25)
```

**Comment by robertjustjones** [comment]
@cbratschi do you have the `connectWithNoPrimary` option enabled?  I've just found that having it enabled breaks with the code change in 9cf1dd4.

**Comment by christkv** [comment]
I would need a standalone code example to diagnose this issue

**Comment by robertjustjones** [comment]
You have a single test `Successful connection to replicaset of 0 primary, 1 secondary and 1 arbiter with secondaryOnlyConnectionAllowed` in test/tests/functional/rs_mocks/connection_tests.js  that sets ` secondaryOnlyConnectionAllowed: true`. 

The equivalent of my case should be to replicate that test with `1 primary, 1 secondary and 1 arbiter`.  Would you mind adding that test in your environment to see if it passes? It's a reasonable test in any case.

**Comment by christkv** [comment]
1 primary, 1 secondary and 1 arbiter with secondaryOnlyConnectionAllowed = true ?

**Comment by christkv** [comment]
@robertjustjones I'm not getting the issue you are seeing with that test.

**Comment by robertjustjones** [comment]
ok, thanks @christkv  I'll dig deeper to try to replicate.

**Comment by cbratschi** [comment]
@robertjustjones yes, it was connectWithNoPrimary set to true in my case. It works now fine again.

**Comment by robertjustjones** [comment]
Great to hear @cbratschi.

@christkv The mapping from connectWithNoPrimary to secondaryOnlyConnectionAllowed to made in https://github.com/mongodb/node-mongodb-native/blob/2.2/lib/utils.js#L245.  Perhaps you could review that code to see if anything other possible triggers come to mind outside of the new test above. 

Also, @cbratschi (I presume) and I are connecting to "real" replicaSets on non-localhost machines. Perhaps there's a race condition that is triggering the error by introducing network delay. Maybe that could be simulated in the test case.

**Comment by christkv** [comment]
There is a fix in 2.2.25 that will fix this, however this is due to using connectWithNoPrimary which should not be used anyway unless you have a workload that only does secondary operations.
